File: lib_common/src/main/java/com/soli/libCommon/base/BaseRecycleAdapter.java
package com.soli.libCommon.base;

import android.content.Context;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;
import java.util.List;

/**
 * 时间：2018/11/13
 * 作者：CDY
 */
public abstract class BaseRecycleAdapter<T> extends RecyclerView.Adapter<RecyclerView.ViewHolder> {

    protected final int ITEM_TYPE_NORMAL = 0;
    protected final int ITEM_TYPE_HEADER = 1;
    protected final int ITEM_TYPE_FOOTER = 2;

    protected List<T> mList;

    protected Context ctx;

    protected View HeaderView, FooterView;

    protected LayoutInflater inflater;

    //更新数据是否要用有动画的那种效果
    protected boolean useHaveAnimationRefresh = true;

    private OnItemClickListener onItemClickListener;

    public BaseRecycleAdapter(Context context) {
        this.ctx = context;
        inflater = LayoutInflater.from(context);
    }

    public BaseRecycleAdapter(Context context, List<T> list) {
        this.ctx = context;
        this.mList = list;
        inflater = LayoutInflater.from(context);
    }


    @NonNull
    @Override
    public RecyclerView.ViewHolder onCreateViewHolder(@NonNull ViewGroup viewGroup, int viewType) {
        return onCreateViewHolder_impl(viewGroup, viewType);
    }

    @Override
    public void onBindViewHolder(RecyclerView.ViewHolder mholder, int position) {

        int type = getItemViewType(position);
        int realPosition = getRealItemPosition(position);

        if (ITEM_TYPE_NORMAL == type) {
            mholder.itemView.setOnClickListener(view -> {
                if (onItemClickListener != null) {
                    onItemClickListener.OnItemClick(realPosition);
                }
            });
        }

        onBindViewHolder_impl(mholder, type, position, realPosition);
    }

    @Override
    public void onBindViewHolder(@NonNull RecyclerView.ViewHolder holder, int position, @NonNull List<Object> payloads) {
        if (payloads.isEmpty())
            super.onBindViewHolder(holder, position, payloads);
        else
            onBindViewHolderPayLoads(holder, getItemViewType(position), position, getRealItemPosition(position), payloads);
    }

    /**
     * @param viewHolder
     * @param itemType
     * @param original_position
     * @param real_position
     * @param payloads
     */
    public void onBindViewHolderPayLoads(RecyclerView.ViewHolder viewHolder, int itemType, int original_position, int real_position, @NonNull List<Object> payloads) {
        //do something
    }

    @Override
    public void onViewDetachedFromWindow(RecyclerView.ViewHolder holder) {
        super.onViewDetachedFromWindow(holder);
        holder.itemView.clearAnimation();
    }


    protected void setText(TextView textView, CharSequence text) {
        if (!TextUtils.isEmpty(text))
            textView.setText(text);
        else
            textView.setText("");
    }

    /**
     * 添加头部
     */
    public void addHeaderView(View view) {
        this.HeaderView = view;
    }

    /**
     * 添加底部
     */
    public void addFooterView(View view) {
        this.FooterView = view;
    }


    /**
     * 获取真正的position,因为加了头部与底部，position会有所位移
     */
    public int getRealItemPosition(int position) {
        if (null != HeaderView) {
            return position - 1;
        }
        return position;
    }

    /**
     * @return
     */
    public int getRealItemCount() {
        return mList == null ? 0 : mList.size();
    }

    @Override
    public int getItemCount() {
        return getRealItemCount() + getHeaderCount() + getFooterCount();
    }

    @Override
    public int getItemViewType(int position) {
        if (HeaderView != null && position == 0) {
            return ITEM_TYPE_HEADER;
        } else if (FooterView != null && getItemCount() - 1 == position) {
            return ITEM_TYPE_FOOTER;
        }
        return ITEM_TYPE_NORMAL;
    }


    /**
     * @param position
     * @return
     */
    public T getItemData(int position) {
        return mList != null && mList.size() > 0 && position < getRealItemCount() ? mList.get(position) : null;
    }

    /**
     * @param list
     */
    public void setList(List<T> list) {
        checkList();
        mList.clear();
        this.mList.addAll(list);
        notifyDataSetChanged();
    }

    public List<T> getList() {
        checkList();
        return this.mList;
    }

    /**
     *
     */
    private void checkList() {
        if (mList == null) {
            mList = new ArrayList<>();
        }
    }

    public void add(T t) {
        checkList();
        mList.add(t);
        if (useHaveAnimationRefresh)
            notifyItemInserted(mList.size() - 1 + getHeaderCount());
        else
            notifyDataSetChanged();
    }

    public int getHeaderCount() {
        return HeaderView != null ? 1 : 0;
    }

    public int getFooterCount() {
        return FooterView != null ? 1 : 0;
    }

    /**
     * insert  a item associated with the specified position of adapter
     *
     * @param position
     * @param item
     */
    public void add(int position, T item) {
        checkList();
        mList.add(position, item);
        if (useHaveAnimationRefresh) {
            notifyItemInserted(position + getHeaderCount());
            notifyItemRangeChanged(position + getHeaderCount(), getItemCount() - position - getHeaderCount());
        } else
            notifyDataSetChanged();
    }


    public void addAll(List<T> newData) {
        if (newData == null)
            return;
        checkList();
        mList.addAll(newData);
        if (useHaveAnimationRefresh) {
            notifyItemRangeInserted(mList.size() - newData.size() + getHeaderCount(), newData.size());
        } else
            notifyDataSetChanged();
    }

    public void addAll_Range(List<T> newData) {
        if (newData == null)
            return;
        checkList();
        mList.addAll(newData);
        notifyItemRangeChanged(mList.size() - newData.size() + getHeaderCount(), newData.size());
    }

    public void addAll(int position, List<T> newData) {
        checkList();
        mList.addAll(position, newData);
        if (useHaveAnimationRefresh) {
            notifyItemRangeInserted(position + getHeaderCount(), newData.size());
            notifyItemRangeChanged(position + getHeaderCount(), getItemCount() - position - getHeaderCount());
        } else
            notifyDataSetChanged();
    }

    public void insertAtTop(T data) {
        insert(0, data);
    }

    /**
     * @param position
     * @param data
     */
    public void insert(int position, T data) {
        checkList();
        mList.add(position, data);
        if (useHaveAnimationRefresh) {
            notifyItemInserted(position + getHeaderCount());
            notifyItemRangeChanged(position + getHeaderCount(), getItemCount() - position - getHeaderCount());
        } else
            notifyDataSetChanged();
    }

    /**
     * @param position
     * @param data
     */
    public void set(int position, T data) {
        if (mList != null && position < getItemCount()) {
            mList.set(position, data);
        }
    }

    /**
     * @param position
     */
    public void remove(int position) {
        try {
            if (mList != null) {
                if (position < mList.size()) {
                    mList.remove(position);
                    if (useHaveAnimationRefresh) {
                        notifyItemRemoved(position + getHeaderCount());
                        notifyItemRangeChanged(position + getHeaderCount(), getItemCount() - position - getHeaderCount());
                    } else
                        notifyDataSetChanged();
                }
            }
        } catch (Exception e) {
            e.printStackTrace();
        }
    }

    public void removeAll() {
        if (mList != null) {
            mList.clear();
            notifyDataSetChanged();
        }
    }

    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public interface OnItemClickListener {
        void OnItemClick(int position);
    }


    protected abstract RecyclerView.ViewHolder onCreateViewHolder_impl(ViewGroup parent, int viewType);

    /**
     * 加了header后，position会有所不同,以下是说明
     *
     * @param itemType          判断是headview，footerView的标识
     * @param original_position 原始的position
     * @param real_position     真正的position，
     */
    protected abstract void onBindViewHolder_impl(RecyclerView.ViewHolder mholder, int itemType, int original_position, int real_position);
}

File: lib_common/src/main/java/com/soli/libcommon/base/BaseRecycleAdapter.kt
package com.soli.libcommon.base

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import androidx.recyclerview.widget.RecyclerView
import androidx.viewbinding.ViewBinding
import com.soli.libcommon.util.clickView
import java.lang.reflect.ParameterizedType

/**
 *
 * @author Soli
 * @Time 2020/4/24 13:42
 */
abstract class BaseRecycleAdapter<T, Binding : ViewBinding>(
    context: Context, list: MutableList<T>?
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    constructor(context: Context) : this(context, null)

    val ITEM_TYPE_NORMAL = 0
    val ITEM_TYPE_HEADER = 1
    val ITEM_TYPE_FOOTER = 2

    private var mList: MutableList<T>? = list

    val ctx: Context = context

    var headerView: View? = null
    var footerView: View? = null

    protected val inflater = LayoutInflater.from(context)

    //更新数据是否要用有动画的那种效果
    var useHaveAnimationRefresh = true

    //item的点击事件
    var onItemClickListener: ((view: View, position: Int, data: T?) -> Unit)? = null


    /***
     *
     */
    private fun initBindView(): Binding {
        val type = javaClass.genericSuperclass as ParameterizedType
        val aClass = type.actualTypeArguments[1] as Class<*>
        val method = aClass.getDeclaredMethod("inflate", LayoutInflater::class.java)
        return method.invoke(null, inflater) as Binding
    }


    override fun onCreateViewHolder(
        viewGroup: ViewGroup, viewType: Int
    ): RecyclerView.ViewHolder {
        return onCreateView(viewGroup, viewType)
    }

    override fun onBindViewHolder(
        mholder: RecyclerView.ViewHolder, position: Int
    ) {
        onBindViewHolder(mholder, position, arrayListOf())
    }

    override fun onBindViewHolder(
        mholder: RecyclerView.ViewHolder, position: Int, payloads: MutableList<Any>
    ) {
        val type = getItemViewType(position)
        val realPosition = getRealItemPosition(position)
        if (ITEM_TYPE_NORMAL == type) {
            onBindView(
                (mholder as BaseViewHolder<Binding>).binding, type, position, realPosition, payloads
            )
            //注意如果 onBindView里面也设置了点击时间，并且这里不为空，那么onBindView里面设置的就会不管用
            if (onItemClickListener != null) {
                mholder.itemView.clickView {
                    onItemClickListener!!(it, realPosition, getItemData(realPosition))
                }
            }
        }
    }

    open fun onCreateView(
        parent: ViewGroup?, viewType: Int
    ): RecyclerView.ViewHolder {
        return BaseViewHolder(initBindView())
    }

    /**
     * 加了header后，position会有所不同,以下是说明
     *
     * @param itemType          判断是headview，footerView的标识
     * @param original_position 原始的position
     * @param real_position     真正的position，
     */
    protected abstract fun onBindView(
        binding: Binding,
        itemType: Int,
        originalPosition: Int,
        realPosition: Int,
        payloads: List<Any>
    )

    override fun onViewDetachedFromWindow(holder: RecyclerView.ViewHolder) {
        super.onViewDetachedFromWindow(holder)
        holder.itemView.clearAnimation()
    }


    /**
     * 获取真正的position,因为加了头部与底部，position会有所位移
     */
    fun getRealItemPosition(position: Int): Int {
        return if (null != headerView) {
            position - 1
        } else position
    }

    /**
     * @return
     */
    val realItemCount: Int
        get() = if (mList == null) 0 else mList!!.size

    override fun getItemCount(): Int {
        return realItemCount + headerCount + footerCount
    }

    override fun getItemViewType(position: Int): Int {
        if (headerView != null && position == 0) {
            return ITEM_TYPE_HEADER
        } else if (footerView != null && itemCount - 1 == position) {
            return ITEM_TYPE_FOOTER
        }
        return ITEM_TYPE_NORMAL
    }

    /**
     * @param position
     * @return
     */
    fun getItemData(position: Int): T? {
        return if (mList != null && mList!!.size > 0 && position < realItemCount) mList!![position] else null
    }

    /**
     * @param list
     */
    var list: MutableList<T>
        get() = checkList()
        set(value) {
            checkList()
            mList!!.clear()
            mList!!.addAll(value!!)
            doSomethingWhenSetList()
            notifyDataSetChanged()
        }

    open fun doSomethingWhenSetList() {}

    /**
     *
     */
    private fun checkList(): MutableList<T> {
        if (mList == null) {
            mList = ArrayList()
        }
        return mList!!
    }

    fun add(t: T) {
        checkList()
        mList!!.add(t)
        if (useHaveAnimationRefresh) notifyItemInserted(mList!!.size - 1 + headerCount) else notifyDataSetChanged()
    }

    val headerCount: Int
        get() = if (headerView != null) 1 else 0

    val footerCount: Int
        get() = if (footerView != null) 1 else 0

    /**
     * insert  a item associated with the specified position of adapter
     *
     * @param position
     * @param item
     */
    fun add(position: Int, item: T) {
        checkList()
        mList!!.add(position, item)
        if (useHaveAnimationRefresh) {
            notifyItemInserted(position + headerCount)
            notifyItemRangeChanged(
                position + headerCount, itemCount - position - headerCount
            )
        } else notifyDataSetChanged()
    }

    fun addAll(newData: List<T>?) {
        if (newData == null || newData.isEmpty()) return
        checkList()
        mList!!.addAll(newData)
        if (useHaveAnimationRefresh) {
            notifyItemRangeInserted(mList!!.size - newData.size + headerCount, newData.size)
        } else notifyDataSetChanged()
    }

    fun addAll_Range(newData: List<T>?) {
        if (newData == null || newData.isEmpty()) return
        checkList()
        mList!!.addAll(newData)
        notifyItemRangeChanged(mList!!.size - newData.size + headerCount, newData.size)
    }

    fun addAll(position: Int, newData: List<T>?) {
        if (newData == null || newData.isEmpty()) return
        checkList()
        mList!!.addAll(position, newData)
        if (useHaveAnimationRefresh) {
            notifyItemRangeInserted(position + headerCount, newData.size)
            notifyItemRangeChanged(
                position + headerCount, itemCount - position - headerCount
            )
        } else notifyDataSetChanged()
    }

    fun insertAtTop(data: T) {
        insert(0, data)
    }

    /**
     * @param position
     * @param data
     */
    fun insert(position: Int, data: T) {
        checkList()
        mList!!.add(position, data)
        if (useHaveAnimationRefresh) {
            notifyItemInserted(position + headerCount)
            notifyItemRangeChanged(
                position + headerCount, itemCount - position - headerCount
            )
        } else notifyDataSetChanged()
    }

    /**
     * @param position
     * @param data
     */
    operator fun set(position: Int, data: T) {
        if (mList != null && position < itemCount) {
            mList!![position] = data
            notifyItemChanged(position)
        }
    }

    /**
     * @param position
     */
    fun remove(position: Int) {
        try {
            if (mList != null) {
                if (position > -1 && position < mList!!.size) {
                    mList!!.removeAt(position)
                    if (useHaveAnimationRefresh) {
                        notifyItemRemoved(position + headerCount)
                        notifyItemRangeChanged(
                            position + headerCount, itemCount - position - headerCount
                        )
                    } else notifyDataSetChanged()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    /**
     * @param position
     */
    fun remove(data: T) {
        try {
            val position = list.indexOf(data)
            if (mList != null) {
                if (position > -1 && position < mList!!.size) {
                    mList!!.removeAt(position)
                    if (useHaveAnimationRefresh) {
                        notifyItemRemoved(position + headerCount)
                        notifyItemRangeChanged(
                            position + headerCount, itemCount - position - headerCount
                        )
                    } else notifyDataSetChanged()
                }
            }
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    open fun removeAll() {
        if (mList != null) {
            mList!!.clear()
            notifyDataSetChanged()
        }
    }
}